src/System.Private.CoreLib/shared/System/Text/Utf8Utility.cs

```diff
@@ -0,0 +1,385 @@
+        }
+
+        /// <summary>
+        /// Given a UTF-8 buffer which has been read into a DWORD in machine endianness,
+        /// returns <see langword="true"/> iff the first two bytes of the buffer match
+        /// the UTF-8 2-byte sequence mask [ 110yyyyy 10xxxxxx ]. This method *does not*
+        /// validate that the sequence is well-formed; the caller must still perform
+        /// overlong form checking.
+        /// </summary>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static bool DWordBeginsWithUtf8TwoByteMask(uint value)
+        {
+            // The code in this method is equivalent to the code
+            // below, but the JITter is able to inline + optimize it
+            // better in release builds.
+            //
+            // if (BitConverter.IsLittleEndian)
+            // {
+            //     const uint mask = 0x0000C0E0U;
+            //     const uint comparand = 0x000080C0U;
+            //     return ((value & mask) == comparand);
+            // }
+            // else
+            // {
+            //     const uint mask = 0xE0C00000U;
+            //     const uint comparand = 0xC0800000U;
+            //     return ((value & mask) == comparand);
+            // }
+
+            return (BitConverter.IsLittleEndian && (((value - 0x000080C0u) & 0x000C0E0u) == 0))
+                || (!BitConverter.IsLittleEndian && (((value - 0xC0800000u) & 0xE0C00000u) == 0));
+        }
+
+        /// <summary>
+        /// Given a UTF-8 buffer which has been read into a DWORD on a little-endian machine,
+        /// returns <see langword="true"/> iff the first two bytes of the buffer are a well-formed
+        /// UTF-8 two-byte sequence. This wraps the mask check and the overlong check into a
+        /// single operation. Returns <see langword="false"/> if running on a big-endian machine.
+        /// </summary>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static bool DWordBeginsWithValidUtf8TwoByteSequenceLittleEndian(uint value)
+        {
+            // Per Table 3-7, valid 2-byte sequences are [ C2..DF ] [ 80..BF ].
+            // In little-endian, that would be represented as:
+            // [ ######## ######## 10xxxxxx 110yyyyy ].
+            // Due to the little-endian representation we can perform a trick by ANDing the low
+            // WORD with the bitmask [ 11000000 11111111 ] and checking that the value is within
+            // the range [ 10000000_11000010, 10000000_11011111 ]. This performs both the
+            // 2-byte-sequence bitmask check and overlong form validation with one comparison.
+
+            Debug.Assert(BitConverter.IsLittleEndian);
+
+            return (BitConverter.IsLittleEndian && UnicodeUtility.IsInRangeInclusive(value & 0xC0FFu, 0x80C2u, 0x80DFu))
+                || (!BitConverter.IsLittleEndian && false); // this line - while weird - helps JITter produce optimal code
```

Comment:
Is this something that can be fixed in the JIT?

cc @dotnet/jit-contrib

Reply:
What code does it produce in the optimal cases? On which platform? How does the slow example look? Which code does it produce?

Reply:
Hmm, this smells like a variation of the `true : false` hack.

Reply:
Yeah, this is similar to #914. The normal version of this code generates something like:
```asm
G_M37426_IG02:
       81E1FFC00000         and      ecx, 0xC0FF
       81C13E7FFFFF         add      ecx, -0x80C2
       83F91D               cmp      ecx, 29
       0F96C0               setbe    al
       0FB6C0               movzx    rax, al
       85C0                 test     eax, eax
       7405                 je       SHORT G_M37426_IG03
       E815FDFFFF           call     System.Console:WriteLine()
G_M37426_IG03:
       B801000000           mov      eax, 42
```
Working on it...